File: BennuLib/IO/NativeFormat.vb
Namespace Bennu.IO
    Public Module NativeFormat
        Public ReadOnly NativeDescriptor As Byte() = {&H1A, &H0D, &H0A, &H0}
        Public Const PaletteSize As Short = 768 ' Size in bytes of a palette
        Public Const ReservedBytesSize As Integer = 576
        <Obsolete>
        Public Const NumberOfControlPointsBitMask As Integer = &HFFF ' Bit mask for the number of control points of
        <Obsolete>
        Public Const AnimationFlagBitMask As Integer = &H1000 ' Bit mask for enabled/disabled animation of

        NotInheritable Class Magic
            ' TODO: Think of a different way to handle this. Does not seem good that we need to
            ' repeat this in the native decoders and in the Magic class
            Public Shared ReadOnly Property ValidFormats As String() = {"map", "m16", "m32", "f16", "fpg", "f32", "fnt", "pal"}

            Public ReadOnly Property FileType As String
            Public ReadOnly Property Version As Integer
            Public ReadOnly Descriptor(4) As Byte

            ' TODO: Check for known types?
            Public Sub New(fileType As String, version As Integer, descriptor() As Byte)
                Me.FileType = fileType.ToLower
                Me.Version = version
                Me.Descriptor = descriptor
            End Sub

            Public ReadOnly Property IsDescriptorValid As Boolean
                Get
                    Return StructuralComparisons.StructuralEqualityComparer.Equals(NativeDescriptor, Descriptor)
                End Get
            End Property

            Public ReadOnly Property Depth As Integer
                Get
                    If Integer.TryParse(FileType.Substring(1, 2), Depth) Then
                        Return Depth
                    Else
                        Return 8
                    End If
                End Get
            End Property

            Public ReadOnly Property IsRecognizedFileType As Boolean
                Get
                    Return Array.IndexOf(ValidFormats, FileType) >= 0
                End Get
            End Property

            Public ReadOnly Property IsValid As Boolean
                Get
                    Return IsRecognizedFileType And IsDescriptorValid
                End Get
            End Property

        End Class

        Structure GlyphInfo
            Public ReadOnly Property Width As Integer
            Public ReadOnly Property Height As Integer
            Public ReadOnly Property YOffset As Integer ' Vertical displacement
            Public ReadOnly Property FileOffset As Integer ' Offset of the graphic in the file

            Public Sub New(width As Integer, height As Integer, yOffset As Integer, fileOffset As Integer)
                Me.Width = width
                Me.Height = height
                Me.YOffset = yOffset
                Me.FileOffset = fileOffset
            End Sub
        End Structure
    End Module
End Namespace

File: BennuLib.Sharp/IO/NativeFormat.cs
using System;
using static System.Collections.StructuralComparisons;

namespace BennuLib.IO
{
	public static class NativeFormat
	{
		public static readonly byte[] Terminator = {
			0x1a,
			0xd,
			0xa,
			0x0
		};
	    
        /// <summary>
        /// The size of the color palette area, in bytes
        /// </summary>
		public const int PaletteSize = 768;

        /// <summary>
        /// The size of the gamma color area, in bytes
        /// </summary>
		public const int ReservedBytesSize = 576;
		
        /// <summary>
        /// Bit mask used to separate the number of pivot points from the animation flags
        /// </summary>
		public const int PivotPointsNumberBitMask = 0xfff;
		
        /// <summary>
        /// Bit mask used to get the animation bit flag. Animation is not supported.
        /// </summary>
		public const int AnimationFlagBitMask = 0x1000;


        /// <summary>
        /// Represents the header of the native formats, i.e. a section describing type
        /// type of file (graphic, graphic collection, font or palettes), and the depth
        /// of the graphic information (1, 8, 16 or 32bpp).
        /// </summary>
		public sealed class Header
		{

			public string Magic { get; }
			public int Version { get; }

			public readonly byte[] Descriptor = new byte[5];
			
			public Header(string magic, int version, byte[] terminator)
			{
				Magic = magic.ToLower();
				Version = version;
				Descriptor = terminator;
			}

			public bool IsTerminatorValid() {
                    return StructuralEqualityComparer.Equals(Terminator, Descriptor);
			}

            /// <summary>
            /// All native format's magic follow the pattern 'aXY' where XY indicates, for
            /// non 8bpp formats, the depth (01, 16 or 32). For 8bpp formats, XY ar two 
            /// characters.
            /// </summary>
			public int Depth
            {
                // TODO: Perhaps it is wiser to have a ParseDepth instead of a property...
                get
                {
                    int depth;

                    if (int.TryParse(Magic.Substring(1, 2), out depth))
                        return depth;
                    else
                        return 8;
                }
            }

		}

		public struct GlyphInfo
		{
			public int Width { get; }
			public int Height { get; }
			public int YOffset { get; }
			// Vertical displacement
			public int FileOffset { get; }
			// Offset of the graphic in the file

			public GlyphInfo(int width, int height, int yOffset, int fileOffset)
			{
				Width = width;
				Height = height;
				YOffset = yOffset;
				FileOffset = fileOffset;
			}
		}
	}
}
